### PR TITLE
[CARBONDATA-2585][CARBONDATA-2586][Local Dictionary]Added test cases for local dictionary support for alter table, set, unset and preaggregate

### PR DESCRIPTION
**Pull request by brijoobopanna**
Dependent on #2401

**What changes were proposed in this pull request?**

In this PR,
Added test cases for local dictionary support for alter table, set, unset and pre-aggregate
All the validations related to above features are taken care in this PR

**How was this patch tested?**
Test case were executed on a 3 node cluster and the same were automated and added as SDV test cases

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6588/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5412/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5481/

**Comment by brijoobopanna** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6606/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6649/

File: integration/spark-common-cluster-test/src/test/scala/org/apache/carbondata/cluster/sdv/generated/CreateTableWithLocalDictionaryTestCase.scala [hunk 1]
```diff
@@ -2096,6 +2096,264 @@ class CreateTableWithLocalDictionaryTestCase extends QueryTest with BeforeAndAft
...
+      sql(
+        "alter table local1 add columns (alt string,abc string) tblproperties" +
+        "('local_dictionary_threshold'='10000')")
+    }
+    assert(exception1.getMessage
+      .contains(
+        "Unsupported Table property in add column: local_dictionary_threshold"))
+  }
+
+  test("test alter table add column when main table is disabled for local dictionary")
```
Comment (kunal642) [inline integration/spark-common-cluster-test/src/test/scala/org/apache/carbondata/cluster/sdv/generated/CreateTableWithLocalDictionaryTestCase.scala]: Add one test case in which 2 string columns are added using alter but only one column is specified in local_dic_include. Both should be considered for local dict generation.

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/localdictionary/LocalDictionarySupportCreateTableTest.scala [hunk 2]
```diff
@@ -2585,67 +2585,445 @@ class LocalDictionarySupportCreateTableTest extends QueryTest with BeforeAndAfte
...
+        | tblproperties('local_dictionary_exclude'='name','local_dictionary_include'='city,dcity',
+        | 'local_dictionary_enable'='true')
+      """.
+        stripMargin)
+    val descFormatted1 = sql("describe formatted local1").collect
+    descFormatted1.find(_.get(0).toString.contains("Local Dictionary Enabled")) match {
+      case Some(row) => assert(row.get(1).toString.contains("true"))
+    }
+    descFormatted1.find(_.get(0).toString.contains("Local Dictionary Include")) match {
+      case Some(row) => assert(row.get(1).toString.contains("city,dcity"))
```
Comment (kunal642) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/localdictionary/LocalDictionarySupportCreateTableTest.scala]: add check to verify that name is not included in local dictionary include property.

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/localdictionary/LocalDictionarySupportCreateTableTest.scala [hunk 3]
```diff
@@ -2585,67 +2585,445 @@ class LocalDictionarySupportCreateTableTest extends QueryTest with BeforeAndAfte
...
+        | tblproperties('local_dictionary_exclude'='st','local_dictionary_include'='city',
+        | 'local_dictionary_enable'='true')
+      """.
+        stripMargin)
+    val descLoc1 = sql("describe formatted local1").collect
+    descLoc1.find(_.get(0).toString.contains("Local Dictionary Include")) match {
+      case Some(row) => assert(row.get(1).toString.contains("city"))
+    }
+    descLoc1.find(_.get(0).toString.contains("Local Dictionary Exclude")) match {
+      case Some(row) => assert(row.get(1).toString.contains("st"))
```
Comment (kunal642) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/localdictionary/LocalDictionarySupportCreateTableTest.scala]: add check for child column also

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5484/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6657/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5519/

File: integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala [hunk 4]
```diff
@@ -278,14 +282,14 @@ class AlterTableColumnSchemaGenerator(
...
       .foreach(f => if (f._2.size > 1) {
-      val name = f._1
-      LOGGER.error(s"Duplicate column found with name: $name")
-      LOGGER.audit(
-        s"Validation failed for Create/Alter Table Operation " +
-        s"for ${ dbName }.${ alterTableModel.tableName }. " +
-        s"Duplicate column found with name: $name")
-      sys.error(s"Duplicate column found with name: $name")
-    })
+        val name = f._1
```
Comment (akashrn5) [inline integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala]: correct the style

File: integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala [hunk 5]
```diff
@@ -299,17 +303,112 @@ class AlterTableColumnSchemaGenerator(
     val columnValidator = CarbonSparkFactory.getCarbonColumnValidator
     columnValidator.validateColumns(allColumns)
 
+
```
Comment (akashrn5) [inline integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala]: add proper comments to all the changes and methods, and you can move this to util class

File: integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala [hunk 6]
```diff
@@ -299,17 +303,112 @@ class AlterTableColumnSchemaGenerator(
...
+
     if (alterTableModel.tableProperties != null) {
       CarbonUtil
         .setLocalDictColumnsToWrapperSchema(newCols.asJava,
           alterTableModel.tableProperties.asJava,
           tableSchema.getTableProperties.get(CarbonCommonConstants.LOCAL_DICTIONARY_ENABLE))
     }
 
+    val newMainList = getLocalDictColumnList(tableSchema.getTableProperties.asScala,
```
Comment (akashrn5) [inline integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala]: give proper variable names, add comments in all places

File: integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala [hunk 7]
```diff
@@ -341,7 +445,7 @@ class AlterTableColumnSchemaGenerator(
           }
         }
         else if (elem._1.equalsIgnoreCase("no_inverted_index") &&
-          (elem._2.split(",").contains(col.getColumnName))) {
+                 (elem._2.split(",").contains(col.getColumnName))) {
```
Comment (akashrn5) [inline integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala]: remove unnecessary changes

File: integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala [hunk 8]
```diff
@@ -318,6 +417,11 @@ class AlterTableColumnSchemaGenerator(
           tablePropertiesMap.put(x._1, x._2)
         }
     }
+
```
Comment (akashrn5) [inline integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala]: mention the reason, why you need to explicitly put these keys into the map

File: integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala [hunk 9]
```diff
@@ -391,7 +495,7 @@ object TableNewProcessor {
     if (dataType == DataTypes.DATE) {
       encoders.add(Encoding.DIRECT_DICTIONARY)
     }
-    if (dataType == DataTypes.TIMESTAMP && ! highCardinalityDims.contains(colName)) {
```
Comment (akashrn5) [inline integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala]: remove this

File: integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala [hunk 10]
```diff
@@ -440,7 +544,7 @@ class TableNewProcessor(cm: TableModel) {
   }
 
   // varchar column is a string column that in long_string_columns
-  private def isVarcharColumn(colName : String): Boolean = {
```
Comment (akashrn5) [inline integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala]: same as above, please crosscheck and remove unnnecessary changes

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6703/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5555/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5530/

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/localdictionary/LocalDictionarySupportAlterTableTest.scala [hunk 11]
```diff
@@ -0,0 +1,1183 @@
...
+      sql(
+        "alter table local1 add columns (alt string) tblproperties('local_dictionary_include'='alt,alt')")
+    }
+    assert(exception.getMessage
+      .contains(
+        "LOCAL_DICTIONARY_INCLUDE/LOCAL_DICTIONARY_EXCLUDE contains Duplicate Columns: alt. " +
+        "Please check the DDL."))
+  }
+
+  test("test alter table add column where duplicate columns present in local dictionary include/exclude")
```
Comment (kunal642) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/localdictionary/LocalDictionarySupportAlterTableTest.scala]: Please correct the test case name.

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/localdictionary/LocalDictionarySupportAlterTableTest.scala [hunk 12]
```diff
@@ -0,0 +1,1183 @@
...
+        "('local_dictionary_exclude'='abc')")
+    }
+    assert(exception2.getMessage
+      .contains(
+        "LOCAL_DICTIONARY_INCLUDE/LOCAL_DICTIONARY_EXCLUDE column: abc does not exist in table. " +
+        "Please check the DDL."))
+  }
+
+  test("test alter table add column for datatype validation")
+  {
```
Comment (kunal642) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/localdictionary/LocalDictionarySupportAlterTableTest.scala]: move opening brackets to previous line. Do this for all test cases

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/localdictionary/LocalDictionarySupportAlterTableTest.scala [hunk 13]
```diff
@@ -0,0 +1,1183 @@
...
+  test("test alter unset for local dictionary _002")
+  {
+    sql("drop table if exists local1")
+    sql(
+      """
+        | CREATE TABLE local1(id int, name string, city string, age int)
+        | STORED BY 'carbondata' tblproperties('local_dictionary_include'='name','local_dictionary_exclude'='city')
+      """.stripMargin)
+
+    sql("alter table local1 unset tblproperties('local_dictionary_exclude')")
```
Comment (kunal642) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/localdictionary/LocalDictionarySupportAlterTableTest.scala]: remove if not needed

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/localdictionary/LocalDictionarySupportAlterTableTest.scala [hunk 14]
```diff
@@ -0,0 +1,1183 @@
...
+    val descLoc1 = sql("describe formatted local1").collect
+    descLoc1.find(_.get(0).toString.contains("Local Dictionary Include")) match {
+      case Some(row) => assert(row.get(1).toString.contains("city"))
+    }
+    sql("alter table local1 set tblproperties('local_dictionary_include'='name')")
+    val descLoc2 = sql("describe formatted local1").collect
+    descLoc2.find(_.get(0).toString.contains("Local Dictionary Include")) match {
+      case Some(row) => assert(row.get(1).toString.contains("name"))
+    }
+    intercept[Exception] {
```
Comment (kunal642) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/localdictionary/LocalDictionarySupportAlterTableTest.scala]: all these negative scenarios are already handled in previous test cases

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/localdictionary/LocalDictionarySupportAlterTableTest.scala [hunk 15]
```diff
@@ -0,0 +1,1183 @@
...
+    descLoc1.find(_.get(0).toString.contains("Local Dictionary Exclude")) match {
+      case Some(row) => assert(row.get(1).toString.contains("city"))
+    }
+    sql("alter table local1 set tblproperties('local_dictionary_exclude'='name')")
+    val descLoc2 = sql("describe formatted local1").collect
+    descLoc2.find(_.get(0).toString.contains("Local Dictionary Exclude")) match {
+      case Some(row) => assert(row.get(1).toString.contains("name"))
+    }
+    intercept[Exception] {
+      sql("alter table local1 set tblproperties('local_dictionary_exclude'='id')")
```
Comment (kunal642) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/localdictionary/LocalDictionarySupportAlterTableTest.scala]: already handled. no need to validate again. Remove the whole test case.

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/localdictionary/LocalDictionarySupportAlterTableTest.scala [hunk 16]
```diff
@@ -0,0 +1,1183 @@
...
+        | CREATE TABLE local1(id int, name string, city string, age int,add string)
+        | STORED BY 'carbondata' tblproperties('local_dictionary_include'='city',
+        | 'local_dictionary_exclude'='name')
+      """.stripMargin)
+
+    val descLoc1 = sql("describe formatted local1").collect
+    descLoc1.find(_.get(0).toString.contains("Local Dictionary Include")) match {
+      case Some(row) => assert(row.get(1).toString.contains("city"))
+    }
+    sql("alter table local1 unset tblproperties('local_dictionary_include')")
```
Comment (kunal642) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/localdictionary/LocalDictionarySupportAlterTableTest.scala]: unset 'local_dictionary_exclude' instead of 'local_dictionary_include' as the name suggests

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/localdictionary/LocalDictionarySupportAlterTableTest.scala [hunk 17]
```diff
@@ -0,0 +1,1183 @@
...
+
+  test("test alter table add column default configs for local dictionary _002") {
+    sql("drop table if exists local1")
+    sql(
+      """
+        | CREATE TABLE local1(id int, name string, city string, age int)
+        | STORED BY 'org.apache.carbondata.format' tblproperties('local_dictionary_enable'='true',
+        | 'local_dictionary_threshold'='20000','no_inverted_index'='name')
+      """.stripMargin)
+    sql("alter table local1 add columns (alt string) " +
```
Comment (kunal642) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/localdictionary/LocalDictionarySupportAlterTableTest.scala]: add columns scenarios are already verified in previous test cases.

File: integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala [hunk 18]
```diff
@@ -278,14 +282,14 @@ class AlterTableColumnSchemaGenerator(
     // Its based on the dimension name and measure name
     allColumns.filter(x => !x.isInvisible).groupBy(_.getColumnName)
       .foreach(f => if (f._2.size > 1) {
```
Comment (kunal642) [inline integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala]: change f =>  to (columnName, count) => for better readability

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6739/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5584/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5572/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6751/

**Comment by praveenmeenakshi56** [comment]
retest this please

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5599/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5583/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6766/

File: integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala (hunk 9 above)
Comment (praveenmeenakshi56) [reply integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala]: these are all handled in #2450

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/localdictionary/LocalDictionarySupportAlterTableTest.scala (hunk 16 above)
Comment (praveenmeenakshi56) [reply integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/localdictionary/LocalDictionarySupportAlterTableTest.scala]: handled in #2450

**Comment by kunal642** [comment]
LGTM
